### ISSUE TITLE
[FEATURE REQUEST] pkgrepo should follow current Debian standards

### ISSUE THREAD

**Issue by OrangeDog** [saltstack/salt#59785]
**Is your feature request related to a problem? Please describe.**
Debian have deprecated `apt-key` and strongly recommended no longer adding third party keys to the global apt trust.
Instead, these recommendations should be followed: https://wiki.debian.org/DebianRepository/UseThirdParty

> The key MUST be downloaded over a secure mechanism like HTTPS to a location only writable by root, which SHOULD be `/usr/share/keyrings`. The key MUST NOT be placed in `/etc/apt/trusted.gpg.d` or loaded by `apt-key add`.
>
> A sources.list entry SHOULD have the signed-by option set. The signed-by entry MUST point to a file, and not a fingerprint.

Note that the name of the key file `NAME-archive-keyring.gpg` and sources list `NAME.list` should match.

**Describe the solution you'd like**
When the `pkgrepo.managed` state should be able (and ideally prefer) to set up repos according to these specifications.
```sls
deb [arch=amd64 signed-by=/usr/share/keyrings/salt-archive-keyring.gpg] http://repo.saltproject.io/py3/debian/10/amd64/latest buster main:
  pkrepo.managed:
    - file: /etc/apt/sources.list.d/salt.list
    - key_url: https://repo.saltproject.io/py3/ubuntu/20.04/amd64/latest/salt-archive-keyring.gpg
```

The same should be done when the key is from a keyserver. The file may also require de-armouring if the repo isn't following the new specification.

To make this easier, the current (inconsistent with other backends) behaviour of requiring the `name` to be the verbatim line may need deprecating or otherwise avoiding. For standards-compliant repos it should be as simple as possible, with `file`, `key_url`, `comps` using the defaults.

```sls
salt-repo:
  pkgrepo.managed:
    - architectures: amd64
    - baseurl: http://repo.saltproject.io/py3/debian/10/amd64/latest  # would ideally default to name
    - dist: {{ grains["oscodename"] }}
```

**Describe alternatives you've considered**
Doing it manually with `file.managed` states and ignoring `pkgrepo`.

**Additional context**
Salt itself echoes this specification in its instructions: https://repo.saltproject.io/#debian
It's a bit paradoxical that it doesn't have the ability to follow it itself.

**Comment by OrangeDog** [comment]
Just noting that an armoured keyring (e.g. `/usr/share/keyrings/salt-archive-keyring.asc`) does also work, though as the standard says:
> The file SHOULD NOT be ascii-armored

**Comment by OrangeDog** [comment]
@sagetherage the label says

> Related to packaging of Salt, not Salt's support for package management.

**Comment by sagetherage** [comment]
true, I likely need a better label, ATM I am not sure what that would be, exactly. perhaps 'package-repo"?

**Comment by sagetherage** [comment]
moving this out of triage as we will need to follow a standard and for debian packages at least we should follow their standard, this will need to be planned, but moving from triage to planning.

**Comment by bryceml** [comment]
Removed from the Packaging project since this is about salt code, and not salt's Packages or Repos directly.

**Comment by mherrb** [comment]
Are there working ways to handle repositories with a separate keyking (signed-by=) in the mean time ? 

Currently for me, when trying to manually add a [signed-by= ] tag to the name attribute, salt keeps adding a new line to the .list file each time it's called.

**Comment by OrangeDog** [comment]
@mherrb 

> Doing it manually with `file.managed` states and ignoring `pkgrepo`.

**Comment by udf2457** [comment]
Do you think its worth adding a mild warning note to the docs ? Just to perhaps point people in the right direction about how to do it properly ? (e.g. present a workaround snippet in the docs ?)

**Comment by eliasp** [comment]
> 
> 
> Just noting that an armoured keyring (e.g. `/usr/share/keyrings/salt-archive-keyring.asc`) does also work, though as the standard says:
> 
> > The file SHOULD NOT be ascii-armored

This refers only to the `*.gpg`-file, which should be a GnuPG binary key, while `*.asc` should be an armored file.
Although the docs on https://repo.saltproject.io/#ubuntu point now towards the binary file, I lean towards defaulting to `.asc`/armored which is "human readable".

**Comment by udf2457** [comment]
@eliasp it clearly says at the top of the first paragraph on the page:

> A binary export (gpg --export) of the key SHOULD be available at the root of the repository under the filename deriv-archive-keyring.gpg, where deriv is the a short name for the repository. The file SHOULD NOT be ascii-armored (gpg --export --armor) although a separate armored version MAY be available under deriv-archive-keyring.asc.

**Comment by eliasp** [comment]
Ok, you convinced me… binary it is then. :smile:

**Comment by Ch3LL** [comment]
I have a WIP PR here: https://github.com/saltstack/salt/pull/61760 . I'm still working on solidifying the code, docs and tests but I wanted to get early feedback on my approach. 

Right now with these changes you would need to update your `pkgrepo.managed` state files to include `- aptkey: False`. I took this approach because I didn't want to change the default behavior. If the cli command `apt-key` does not exist it will automatically set `aptkey` to False though in preparation for when `apt-key` is actually removed. Any opposition to this approach?

**Comment by OrangeDog** [comment]
> you would want to change your keydir

But you wouldn't want to change it to `/etc/apt/trusted.gpg.d`. Only the disto maintainers should be putting anything in there.

**Comment by udf2457** [comment]
@Ch3LL , I second what @OrangeDog says, infact the docs are extremely clear in that regard 👍 

> The key MUST NOT be placed in /etc/apt/trusted.gpg.d 

(The CAPS are the docs emphasis, not mine)

**Comment by Ch3LL** [comment]
K, I won't include mention of that directory in my docs. Thanks

**Comment by smcv** [comment]
> Edited: I also set the default `keydir` to save the keys in `/usr/share/keyrings`. This is the default location to save keys for third party keys

The apt 2.4.0 changelog includes this:

>   * Install an empty /etc/apt/keyrings directory.
>     This directory is intended to provide an alternative to
>     /usr/share/keyrings for placing keys used with signed-by.

`/usr` is jointly "owned" by all the apt/dpkg packages that are installed on the system), so `/etc/apt/keyrings` might be a more appropriate location than `/usr/share/keyrings` for keys that are installed by sysadmin action (with Salt or by hand).

There's nothing special about `/usr/share/keyrings` or `/etc/apt/keyrings` specifically, so `/etc/apt/keyrings` should be equally fine to use with older versions of apt, as long as Salt creates it with `mkdir -p` or equivalent before writing into it.

The important thing here is that Salt **doesn't** write into `/etc/apt/trusted.gpg.d`, because keyrings in `/etc/apt/trusted.gpg.d` are trusted to authenticate **all** apt repositories, including the official Debian repository.

**Comment by udf2457** [comment]
@smcv "might be a more appropriate location than /usr/share/keyrings"

Huh ?

The Debian docs are pretty black & white on the topic.....

> The key MUST be downloaded over a secure mechanism like HTTPS to a location only writable by root, which SHOULD be /usr/share/keyrings

(Debian's caps .... not mine !)

https://wiki.debian.org/DebianRepository/UseThirdParty

**Comment by smcv** [comment]
I've asked for clarification in `#debian-apt`.

> you wouldn't want to change it to /etc/apt/trusted.gpg.d. Only the disto maintainers should be putting anything in there.

That's the really important bit: `/etc/apt/trusted.gpg.d` vs. "not `/etc/apt/trusted.gpg.d`".

**Comment by udf2457** [comment]
Arguably the location of storage is minor, the `signed-by=....` in the repo conf file is the more important one !

**Comment by smcv** [comment]
> I'll check with the apt maintainers what their intention was, and if it conflicts with the wiki page, I'll edit the wiki page to match.

Done. It now says the key SHOULD be in `/usr/share/keyrings` if its updates will be managed by an apt/dpkg package like `deriv-archive-keyring`, or in `/etc/apt/keyrings` if it's locally-managed (with Salt and its competitors being tools to help a sysadmin to automate locally-managed stuff).

**Comment by heini** [comment]
Yes, I know. But why not support NOW what is already known to be coming?

**Comment by udf2457** [comment]
@heini Erm, becuase everybody that has installed the current stable version, i.e. bullseye, will not have `/etc/apt/keyrings`.  And I'm not sure `salt` should be doing the job of creating `/etc/apt/keyrings`.  

`salt` should be intelligent enough to determine if it is on bullseye.

**Comment by heini** [comment]
> And I'm not sure `salt` should be doing the job of creating `/etc/apt/keyrings`

Why not? Salt is all about managing files and directories everywhere in the system...

**Comment by dhs-rec** [comment]
BTW: Everyone refers to https://wiki.debian.org/DebianRepository/UseThirdParty, and then seems to take /usr/share/keyrings as the way to go. But it clearly says:
> If future updates to the key will be managed by an apt/dpkg package as recommended below, then it SHOULD be downloaded into /usr/share/keyrings using the same filename that will be provided by the package. If it will be managed locally, it SHOULD be downloaded into /etc/apt/keyrings instead.

I guess the latter is what we do with Salt states. I just know of a few projects that provide keyring packages. In addition, right below this there's also:
> In releases older than Debian 12 and Ubuntu 22.04, /etc/apt/keyrings does not exist by default. It SHOULD be created with permissions 0755 if it is needed and does not already exist.

**Comment by Ch3LL** [comment]
In my PR here: https://github.com/saltstack/salt/pull/61760/files#diff-4942e96e878c10bcad53f5ade902cb904c79835fc642e8a476876afeff18009fR2305 

I am using /usr/share/keyrings by default. Based off this conversation I think I just need to add logic to determine what version of apt we are using and instead use `/etc/apt/keyrings` if  apt >= 2.4.0 instead by default. Is that a correct summary of this discussion?

**Comment by lmf-mx** [comment]
An important consideration is the future behavior of updates of the packages provided by the repo managed with this method.

Should the package be allowed to update and **updates the key (as keys do expire, get compromised)** and it does so with the expectation that the key should live in `/usr/share/keyrings` by following the same standard for a package managing a key.... you now have 2 keys deployed to a system for the same repo.

I, personally would defer to using the same location the package would expect to find a key to update. Others will probably want the separation. The only way to handle all scenarios might be to add an another optional arg with a default path.

**Comment by udf2457** [comment]
@Ch3LL 

TL;DR:  Personally yes, I would agree with that summary.

Reasoning:

Debian Bullseye is still fairly "new" (released 2021-08-14) therefore most people will only "just" be moving over to it and will be using it for sometime. 

Therefore it is reasonable to expect that user base to be using the <2.4.0 apt location of `/usr/share/keyring`.   In other words they will be using `/usr/share/keyring` for "other things" and so it is only fair and reasonable that `salt` should put stuff in the same place so that Bullseye sysadmins only have one place to look.

**Comment by dhs-rec** [comment]
Sorry, but no. Again: Everyone mentions https://wiki.debian.org/DebianRepository/UseThirdParty. That document says: If key is also managed by a package => /usr/share/keyrings, otherwise => /etc/apt/keyrings, regardless of apt version:
> In releases older than Debian 12 and Ubuntu 22.04, /etc/apt/keyrings does not exist by default. It SHOULD be created with permissions 0755 if it is needed and does not already exist.

**Comment by udf2457** [comment]
How about you go get Debian to update the man page for Bullseye and then I'll agree with you ?

**Comment by dhs-rec** [comment]
According to that man-page, which also clearly says (right at the top) that this utility is deprecated, neither /usr/share/keyrings, nor /etc/apt/keyrings should be used, but /etc/apt/trusted.gpg.d:
> /etc/apt/trusted.gpg.d/
File fragments for the trusted keys, additional keyrings can be stored here (by other packages or the administrator). Configuration Item Dir::Etc::TrustedParts.

Which is (again) in contrast to the wiki page, which says:
> The key MUST NOT be placed in /etc/apt/trusted.gpg.d or loaded by apt-key add.

The question now is whether to follow a man-page of a deprecated tool or follow the Debian guidelines...

**Comment by udf2457** [comment]
I already told you previously that you linked to the `testing` release man page, not the `bullseye` manpage.

**Comment by dhs-rec** [comment]
But now I was referring to the bullseye man-page, as linked by you: https://manpages.debian.org/bullseye/apt/apt-key.8.en.html
> apt-key - Deprecated APT key management utility

**Comment by Ch3LL** [comment]
I merged in https://github.com/saltstack/salt/pull/61760 but I will follow up next week with another PR with these changes.

**Comment by dhs-rec** [comment]
It's still not quite correct, sorry. It says (in pkgrepo.py):
> The recommended way to manage repo keys going forward is to download the keys into /etc/apt/keyrings...

But, again, this is only half of the truth. The wiki page states (see above):
> If future updates to the key will be managed by an apt/dpkg package as recommended below, then it SHOULD be downloaded into /usr/share/keyrings using the same filename that will be provided by the package. If it will be managed locally, it SHOULD be downloaded into /etc/apt/keyrings instead.

So the destination directory really depends on how the specific key will be managed after its initial creation...

**Comment by Ch3LL** [comment]
Since we are using the pkgrepo state to manage the key and not a package it would be /etc/apt/keyrings then. Thats how I understand that statement.

**Comment by OrangeDog** [comment]
@dhs-rec there's no possible way for Salt to know the future plans of some other developer. If the user happens to know that it will be package-managed in future, then they can pass `/usr/share/keyrings` to the state. The default should be `/etc/apt/keyrings`.

**Comment by heini** [comment]
But the person writing the Salt state knows. There are projects out there which provide keyring packages. In this case one would simply:
```
foo:
  pkgrepo.managed:
    - humanname: Foo Package Repository
      ...
      require_in:
        - pkg: foo
        - pkg: foo-keyring
  pkg.installed:
    - version: latest
foo-keyring:
  pkg.installed:
    - version: latest
```
In this case the key should go into /usr/share/keyrings, otherwise into /etc/apt/keyrings.

**Comment by Ch3LL** [comment]
As @OrangeDog stated you would need to pass in the keydir path. If you are using the `pkgrepo.managed` state the signedby argument needs to be passed in via the name of the state or the signedby state arg or else the repo will not know where the key is.

**Comment by heini** [comment]
Yes, sure.

**Comment by Ch3LL** [comment]
Closing now that https://github.com/saltstack/salt/pull/62159 is merged

**Comment by OrangeDog** [comment]
> pass in the keydir path

> the signedby state arg

Neither of these have been documented.

**Comment by Ch3LL** [comment]
The signedby state arg was removed here: https://github.com/saltstack/salt/pull/62215 with an explanation in the PR.
The keydir argument is documented here: https://docs.saltproject.io/en/latest/ref/modules/all/salt.modules.aptpkg.html#salt.modules.aptpkg.add_repo_key and on all of the methods that accept that argument.

**Comment by OrangeDog** [comment]
Oh I see. Thought it was going to be on the state function, given the previous discussion.

**Comment by udf2457** [comment]
I'm with @OrangeDog .  Bit of a weird decision.

**Comment by heini** [comment]
I wonder whether it would make sense to simply write the files in `deb822` format (.sources instead of .list), which would allow for a more 1:1 like relation between options of this resource and those in the resulting file. For example:
```
deb http://deb.debian.org/debian bullseye main contrib non-free
deb-src http://deb.debian.org/debian bullseye main contrib non-free
deb http://deb.debian.org/debian bullseye-updates main contrib non-free
deb-src http://deb.debian.org/debian bullseye-updates main contrib non-free
```
would become
```
Types: deb deb-src
URIs: http://deb.debian.org/debian
Suites: bullseye bullseye-updates
Components: main contrib non-free
```
Options in `[]`, like `signed-by` or `arch` simply become an additional entry, like
```
Architectures: amd64 arm64
Signed-By: /path/to/keyring
```
See `sources.list(5)` for the details.

**Comment by OrangeDog** [comment]
It should also be possible to support both formats without breaking anything. For example, if the `name` starts with `deb` or if `file` ends with `.list` then use list format, otherwise use sources format.

**Comment by jdelic** [comment]
hmm, I would prefer an explicit flag (defaulting to whatever fits backwards compatibility) over magic any day.
